### PR TITLE
Implement primary replica side leaseGrant handler

### PR DESCRIPTION
**Pull request by denis-chudov**
https://issues.apache.org/jira/browse/IGNITE-18744

### CODE REVIEW

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java [hunk 1]
```diff
@@ -69,4 +121,96 @@ public CompletableFuture<?> processRequest(ReplicaRequest request) {
...
+     */
+    public CompletableFuture<LeaseGrantedMessageResponse> processLeaseGrantedMessage(LeaseGrantedMessage msg) {
+        return leaderFuture().thenCompose(leader -> {
+            if (hasAcceptedLease(msg.leaseStartTime(), msg.leaseExpirationTime())) {
+                return acceptLease(msg.leaseStartTime(), msg.leaseExpirationTime());
+            } else if (msg.force()) {
+                if (!leader.equals(localNodeSupplier.get())) {
+                    // Replica must wait till safe time reaches the lease start timestamp to make sure that all updates made on the
+                    // group leader are received.
+                    return safeTime.waitFor(msg.leaseStartTime()).thenCompose(v -> {
```
Comment (vldpyatkov) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java:166]: I remember we consider a safe time as a time on leader. But in the new reality it is a time in primary replica, possibly.
If the safe time is replicated from the primary replica, we have no opportunity to use it for primary replica choosing.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java [hunk 2]
```diff
@@ -17,33 +17,85 @@
...
+            ReplicaListener listener,
+            PendingComparableValuesTracker<HybridTimestamp> safeTime,
+            TopologyAwareRaftGroupService raftClient,
+            Supplier<ClusterNode> localNodeSupplier
     ) {
         this.replicaGrpId = replicaGrpId;
         this.listener = listener;
+        this.safeTime = safeTime;
+        this.raftClient = raftClient;
+        this.localNodeSupplier = localNodeSupplier;
```
Comment (vldpyatkov) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java:96]: Is it really so complicated to use a supplier for the local node?

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 3]
```diff
@@ -106,101 +119,140 @@ public ReplicaManager(
         this.clusterNetSvc = clusterNetSvc;
         this.clock = clock;
         this.messageGroupsToHandle = messageGroupsToHandle;
-        this.handler = (message, senderConsistentId, correlationId) -> {
-            if (!busyLock.enterBusy()) {
-                throw new IgniteException(new NodeStoppingException());
+        this.localNodeSupplier = () -> clusterNetSvc.topologyService().localMember();
+        this.handler = this::onReplicaMessageReceived;
+        this.placementDriverMessageHandler = this::onPlacementDriverMessageReceived;
```
Comment (vldpyatkov) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:124]: Why do you save the link if it is just a reference to the local method?

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 4]
```diff
@@ -106,101 +119,140 @@ public ReplicaManager(
...
-                });
-            } finally {
-                busyLock.leaveBusy();
+    private void onPlacementDriverMessageReceived(NetworkMessage msg0, String senderConsistentId, @Nullable Long correlationId) {
+        if (!busyLock.enterBusy()) {
+            throw new IgniteException(new NodeStoppingException());
+        }
+
+        try {
+            if (!(msg0 instanceof PlacementDriverReplicaMessage)) {
```
Comment (vldpyatkov) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:208]: Which other types of messages do you wanting here? I doubt the other type is possible.
It this case, assert is more preferable.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 5]
```diff
@@ -871,7 +878,7 @@ private void updateAssignmentInternal(ConfigurationNotificationEvent<byte[]> ass
                         .thenCompose(v -> storageUpdateHandlerFut)
                         .thenComposeAsync(v -> {
                             try {
-                                return raftMgr.startRaftGroupService(replicaGrpId, newConfiguration);
+                                return raftMgr.startRaftGroupService(replicaGrpId, newConfiguration, raftGroupServiceFactory);
```
Comment (vldpyatkov) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:881]: The same as before, too complicated to hold a factory only to pass it to method start.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java [hunk 6]
```diff
@@ -69,4 +121,96 @@ public CompletableFuture<?> processRequest(ReplicaRequest request) {
...
+     *
+     * @param msg Message to process.
+     * @return Future that contains a result.
+     */
+    public CompletableFuture<LeaseGrantedMessageResponse> processLeaseGrantedMessage(LeaseGrantedMessage msg) {
+        return leaderFuture().thenCompose(leader -> {
+            if (hasAcceptedLease(msg.leaseStartTime(), msg.leaseExpirationTime())) {
+                return acceptLease(msg.leaseStartTime(), msg.leaseExpirationTime());
+            } else if (msg.force()) {
+                if (!leader.equals(localNodeSupplier.get())) {
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java:163]: this is unlikely situation but it doesnt break anything, as the leader can be not colocated with the primary

**Pull request merged**